src/commands/join.cpp

```diff
@@ -0,0 +1,160 @@
+#include "../includes/serverConfig.hpp"
+#include "../includes/utils.hpp"
+
+void Server::joinChannel(const std::string &param, int fd) {
+  // Check if the user is attempting to leave all channels (using "0" as param)
+  if (isLeaveAllChannelsRequest(param)) {
+    // partAllChannels(fd); // Leave all channels function
+    return;
+  }
+
+  Client client = getClientByFd(fd);
+  std::string::size_type spacePos = param.find(" ");
+  std::string channelsPart = param.substr(0, spacePos);
+  std::string keysPart =
+      (spacePos != std::string::npos) ? param.substr(spacePos + 1) : "";
+  stringVector channels;
+  splitByComma(channelsPart, &channels);
+  stringVector keys;
+  splitByComma(keysPart, &keys);
+
+  if (param.empty() || (param.length() == 1 && param[0] == REG_CHAN)) {
+    send461NeedMoreParams(client, "JOIN");
+    return;
+  }
+
+  for (size_t i = 0; i < channels.size(); ++i) {
+    std::string channelName = channels[i];
+
+    if (channelName.empty() ||
+        (channelName.length() == 1 && channelName[0] == REG_CHAN)) {
+      send461NeedMoreParams(client, "JOIN");
+      continue;
+    }
+
+    std::string channelNameWithoutPrefix = channelName.substr(1);
+
+    if (!isChannelValid(channelName, client)) {
+      continue;
+    }
+
+    if (_channels.find(channelNameWithoutPrefix) == _channels.end()) {
+      createAndRegisterChannel(&client, channelNameWithoutPrefix);
+    }
+
+    std::string key = (i < keys.size()) ? keys[i] : "";
+    Channel &channel = _channels[channelNameWithoutPrefix];
+
+    if (channel.getMode().keyRequired && key != channel.getKey()) {
+      send475BadChannelKey(client, channel);
+      continue;
+    }
+
+    handleJoinRequest(fd, client, channelNameWithoutPrefix);
+  }
+}
+
+bool Server::isValidChannelNameLength(const std::string &param) {
+  return param.length() >= 1 && param.length() < 51;
+}
+
+bool Server::isValidChannelPrefix(const std::string &param) {
+  return param[0] == REG_CHAN;
+}
+
+bool Server::isSingleCharacterChannelPrefix(const std::string &param) {
+  return param.length() == 1 && param[0] == REG_CHAN;
+}
+
+bool Server::isLeaveAllChannelsRequest(const std::string &param) {
+  return param == "0";
+}
+
+bool Server::isChannelValid(const std::string &param, const Client &client) {
+  std::cout << " param Length: " << param.length() << std::endl;
+  if (!isValidChannelNameLength(param) || !isValidChannelPrefix(param)) {
+    send476BadChanMask(client, param);
+    return false;
+  } else if (static_cast<size_t>(client.getChannelsCount()) >=
+             gConfig->getLimit("CHANLIMIT")) {
+    send405TooManyChannels(client);
+    return false;
+  }
+  return true;
+}
+
+void Server::createAndRegisterChannel(Client *client,
+                                      const std::string &channelName) {
+  addChanneltoServerIfNoExist(channelName);
+  _channels[channelName].addClientToChannelMap(client);
+  _channels[channelName].addOperatorsToChannelMap(client);
```

Comment:
**We should use `void addOperator(Client *client);` defined at line 194 on Channel.cpp.**   
--> `_channels[channelName].addOperator(client);`

Reason:
The client has to be added to `clientPMap _channelOperators` rather than `clientPMap _channelClients;`. He has already been added to `clientPMap _channelClients;` by `_channels[channelName].addClientToChannelMap(client);` (at line 110).

Reply:
the functoin addOperatorsToChannelMap was corrected but duplicated with your function addOperator, i have removed it.